onnxruntime/core/session/inference_session.cc

```diff
@@ -1464,12 +1464,16 @@ common::Status InferenceSession::ValidateInputs(const std::vector<std::string>&
     auto expected_type = iter->second.ml_data_type;
     auto& input_ml_value = feeds.at(i);
     if (input_ml_value.IsTensor()) {
-      // check for type
-      if (!expected_type->IsTensorType()) {
+      if (!expected_type->IsTensorType() &&
+          !(expected_type->IsOptionalType() &&
```

Comment:
Legal to receive tensors as long as the expected input is a `tensor` or an `optional(tensor)`